app/services/activitypub/fetch_replies_service.rb

```diff
@@ -0,0 +1,51 @@
+# frozen_string_literal: true
+
+class ActivityPub::FetchRepliesService < BaseService
+  include JsonLdHelper
+
+  def call(parent_status, collection_or_uri)
+    @account = parent_status.account
+
+    @items = collection_items(collection_or_uri)
+
+    if @items.nil?
+      uri = value_or_id(collection_or_uri)
+      return if invalid_origin?(uri)
+      collection = fetch_resource_without_id_validation(uri)
+      raise Mastodon::UnexpectedResponseError if collection.nil?
+      @items = collection_items(collection)
+      return if @items.nil?
+    end
+
+    FetchReplyWorker.push_bulk(filtered_replies)
+  end
+
+  private
+
+  def collection_items(collection)
+    return unless collection.is_a?(Hash)
+    case collection['type']
+    when 'Collection', 'CollectionPage'
+      collection['items']
+    when 'OrderedCollection', 'OrderedCollectionPage'
+      collection['orderedItems']
+    end
+  end
+
+  def filtered_replies
+    # Only fetch replies to the same server as the original status to avoid
+    # amplification attacks.
+
+    # Also limit to 5 fetched replies to limit potential for DoS.
+    @items.map { |item| value_or_id(item) }.reject { |uri| invalid_origin?(uri) }.take(5)
```

Comment:
Same comment here. I understand the DoS concern, and I don't know what to do here. 5 seems unrealistically low though. Perhaps I should run a SQL query to check average self-replies length on mastodon.social?

Reply:
For the purpose of thread-fetching, I think 5 is fine, as long as immediate self-replies are listed first. Indeed, if we take https://mastodon.technology/@ashfurrow/101341748880807263, the first status will list https://mastodon.technology/@ashfurrow/101358323548336310, which will itself list https://mastodon.technology/@ashfurrow/101358324411505893 and https://mastodon.technology/@ashfurrow/101364619725098480, and so on.